### PR TITLE
Add cppstd as a subsetting of compiler

### LINKED ISSUE #4873: cppstd as a subsetting
We considered better to have the cppstd as a subsetting of every compiler:

- If you don't specify the -s compiler.cppstd=XXX it won't change anything, because will come with a None default value.
- We won't break anything, the global setting will be kept, but deprecated (comment at settings.yml and docs).
- So any recipe could receive the subsetting and compile with a different version of the language automatically.
- The reasons to do a first level setting was to not generate new binary IDS for C libraries that doesn't remove the subsetting. We consider more important to be "injectable" at any recipe.
- Conan will fail if both are specified.
- Conan will try internally to manage the value of the new one, copying from the old one if possible (to ease the code and future deprecation of the global setting)
- Each compiler will have their own values, it has a lot of sense.

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Feature: Add `compiler.cppstd` setting (mark `cppstd` as deprecated)
Docs: https://github.com/conan-io/docs/pull/1266

This PR includes changes from PR #4986 and #4942 (really short ones)

----

Some hints about the new/old behavior:
 - It keeps previous behavior: all use cases (tests) should be summarized here https://github.com/conan-io/conan/pull/4986, 
 - The default value (or `None`) for `compiler.cppstd` generates the same ID as it was generating for `cppstd`.
 - If the user specifies `cppstd` and `compiler.cppstd`, Conan will raise. Yes, even if those have the same values.
 - Package ID for non-default values generates different ID for `cppstd` and `compiler.cppstd`
 - If no value is provided to `compiler.cppstd` then the default value will be assigned to it.

Summarizing:
```
package_id("cppstd=default") == package_id("cppstd=None")
package_id("cppstd=default") == package_id("compiler.cppstd=default")
package_id("compiler.cppstd=default") == package_id("compiler.cppstd=None")
package_id("cppstd=20") != package_id("compiler.cppstd=20")
```

----

closes #4873 

@TAGS: slow

### CODE REVIEW

File: conans/model/info.py [hunk 1]
```diff
@@ -397,15 +397,16 @@ def default_std_matching(self):
         If we are building with gcc 7, and we specify -s cppstd=gnu14, it's the default, so the
         same as specifying None, packages are the same
         """
-
-        if self.full_settings.cppstd and \
-                self.full_settings.compiler and \
-                self.full_settings.compiler.version:
+        if self.full_settings.compiler and \
```
Comment (jgsogo) [inline conans/model/info.py]: Do we want to maintain this `default` behavior?

File: conans/test/functional/configuration/compiler_cppstd_test.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    def test_value_different_with_general_setting(self):
+        with catch_deprecation_warning(self, n=2):
+            self.t.run("create . hh/0.1@user/channel"
+                       " -s cppstd=11"
+                       " -s hh:compiler=apple-clang"
+                       " -s hh:compiler.cppstd=14", assert_error=True)
+        self.assertIn("Package 'hh': The specified 'compiler.cppstd=14' and 'cppstd=11'"
+                      " are different", self.t.out)
+
+    @unittest.expectedFailure
```
Comment (jgsogo) [inline conans/test/functional/configuration/compiler_cppstd_test.py]: In this test, although the recipe does not define `settings = ... "compiler"`, the settings object contains that value, it detects the mismatch and raises.

**Comment by jgsogo** [comment]
Green!

image

File: conans/model/conan_file.py [hunk 3]
```diff
@@ -145,6 +152,13 @@ def initialize(self, settings, env):
         except ConanException:
             pass
 
+        try:
```
Comment (jgsogo) [inline conans/model/conan_file.py]: Need to try/except?

File: conans/model/conan_file.py [hunk 4]
```diff
@@ -132,6 +133,12 @@ def initialize(self, settings, env):
         self.options = create_options(self)
         self.requires = create_requirements(self)
         self.settings = create_settings(self, settings)
+
+        try:
```
Comment (memsharded) [inline conans/model/conan_file.py]: I think this check might be redundant:
- It is not possible to get cppstd defined if it doesn't come from the profile
- The profile was checked before being applied to conanfiles

If the problem is per-package settings, I think I'd try to implement that in the same place as the global cppstd is checked, to keep that logic together.

File: conans/test/functional/cppstd/default_cppstd_test.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                self.output.info(">> cppstd: {{}}".format(cppstd))
+                self.output.info(">> compiler.cppstd: {{}}".format(compiler_cppstd))
+        """)
+
+    id_default = "d17189cfe7b11efbc5d701339a32d203745f8b81"
+
+    def run(self, *args, **kwargs):
+        default_profile_path = os.path.join(temp_folder(), "default.profile")
+        save(default_profile_path, self.default_profile)
+        with environment_append({"CONAN_DEFAULT_PROFILE_PATH": default_profile_path}):
```
Comment (memsharded) [inline conans/test/functional/cppstd/default_cppstd_test.py]: This looks complicated, why not directly writing the home/profiles/default profile?

File: conans/model/conan_file.py (hunk 4 above)
Comment (jgsogo) [reply conans/model/conan_file.py]: Yes, this is about checking the _per-package settings_. 

I tried to do it (there are some ideas in [this PR](https://github.com/jgsogo/conan/pull/11) which I moved from this repo to mine), but there is a problem: **I need to constrain the settings for the package before checking them** and that action is performed inside this function, in the call to `self.settings = create_settings(self, settings)`.

I would really like to move this check to previous stages (this is executed once the graph starts to run), but it would require to move too many things not related to this PR, don't you think so?

File: conans/test/functional/cppstd/default_cppstd_test.py (hunk 5 above)
Comment (jgsogo) [reply conans/test/functional/cppstd/default_cppstd_test.py]: Why? It is a documented way to override the default profile: _"[TIP:] using the environment variable `CONAN_DEFAULT_PROFILE_PATH`"_. I think this is the easiest way to be sure that every `TestClient` created in these tests will use the same profile...

File: conans/client/settings_preprocessor.py [hunk 6]
```diff
@@ -1,30 +1,73 @@
...
-                                                                       compiler_version,
-                                                                       available))
+
+    # Checks: one or the other, but not both
+    if cppstd and compiler_cppstd:
+        raise ConanException("Do not use settings 'compiler.cppstd' together with 'cppstd'."
+                             " Use only the former one.")
+
+    if cppstd:
+        warnings.warn("Setting 'cppstd' is deprecated in favor of 'compiler.cppstd'")
```
Comment (uilianries) [inline conans/client/settings_preprocessor.py]: I'm curious, why not `self.output.warn`?

File: conans/client/migrations.py [hunk 7]
```diff
@@ -154,6 +154,8 @@ def _make_migrations(self, old_version):
         if old_version < Version("1.15.0"):
             migrate_registry_file(self.cache, self.out)
 
+        # TODO: Add settings migration
```
Comment (lasote) [inline conans/client/migrations.py]: I don't think we should. The current policy is just to fast forward if not changed. Related with #5038

File: conans/client/settings_preprocessor.py (hunk 6 above)
Comment (jgsogo) [reply conans/client/settings_preprocessor.py]: This raises a `UserWarning` if you have configured _warnings as errors_ (we do in the test suite), if not it just prints the message.

File: conans/model/conan_file.py (hunk 4 above)
Comment (lasote) [reply conans/model/conan_file.py]: I don't get it. I will go ask you "live" LoL

File: conans/model/info.py [hunk 8]
```diff
@@ -398,14 +401,36 @@ def default_std_matching(self):
...
+                        return settings
+                self._std_matching = remove_cppstd
+
     def default_std_non_matching(self):
         if self.full_settings.cppstd:
             self.settings.cppstd = self.full_settings.cppstd
+        self._std_matching = None  # Do nothing
+
+    # Handle std_matching strategy for compiler.cppstd
+    _std_matching = None
```
Comment (lasote) [inline conans/model/info.py]: This variable here feels weird, why is it needed here and not at `__init__`?

Comment (lasote) [reply conans/model/info.py]: Maybe it reads easier if the var is always a function (empty lambda by default?) and it is called differently, something like: `self._adjust_settings_for_std`. You wouldn't need to write `if self._std_matching` just call always the `self._adjust_settings_for_std`.

File: conans/test/functional/configuration/profile_test.py [hunk 9]
```diff
@@ -551,6 +551,7 @@ def profile_crazy_inheritance_test(self):
                              [settings]
                              arch=x86_64
                              compiler=Visual Studio
+                             compiler.cppstd=14
```
Comment (lasote) [inline conans/test/functional/configuration/profile_test.py]: Why this change?

File: conans/test/functional/cppstd/default_cppstd_test.py [hunk 10]
```diff
@@ -34,14 +33,15 @@ class Library(ConanFile):
 
             def configure(self):
                 cppstd = self.settings.get_safe("cppstd")
-                self.output.info(">>>> settings: {{}}".format(self.settings.fields))
-                self.output.info(">>>> cppstd: {{}}".format(cppstd))
+                compiler_cppstd = self.settings.get_safe("compiler.cppstd")
```
Comment (lasote) [inline conans/test/functional/cppstd/default_cppstd_test.py]: Do we want to alter this test? Or keep it to ensure we don't break anything and do another one for the new subsetting?

Comment (jgsogo) [reply conans/test/functional/cppstd/default_cppstd_test.py]: I've minimized changes in this file, now I think it is clear that the modifications (additions) don't affect existing behavior. Tell me what you think, thanks!

File: conans/test/functional/configuration/profile_test.py (hunk 9 above)
Comment (jgsogo) [reply conans/test/functional/configuration/profile_test.py]: Now we are setting `compiler.cppstd` to the default value if the user is not providing one (check function `_fill_compiler_cppstd`) although we remove it for ID computation.

File: conans/client/migrations.py (hunk 7 above)
Comment (jgsogo) [reply conans/client/migrations.py]: Understood, didn't know about that issue/PR.

File: conans/model/info.py [hunk 11]
```diff
@@ -398,14 +404,31 @@ def default_std_matching(self):
...
+                        return settings
+                self._adjust_settings_for_std = remove_cppstd
+
     def default_std_non_matching(self):
         if self.full_settings.cppstd:
             self.settings.cppstd = self.full_settings.cppstd
+        self._adjust_settings_for_std = lambda u: u  # Do nothing
+
+    def _settings_sha(self):
+        settings = self.settings.copy()
```
Comment (memsharded) [inline conans/model/info.py]: I suggest:
- Having a default adjust = None
- If adjust = None, no need to copy, no need to call it.
- The adjust would do the copy internally, and return the sha. In this way it is not possible to forget to copy, which would destroy the ``info`` original values.

Comment (jgsogo) [reply conans/model/info.py]: I agree with the `adjust=None`, but I prefer to leave the `copy` and `sha` outside the _adjust_ function, maybe we can add here more _adjustments_ only for computing the ID and this way we can create the pipeline (I'm not sure if it will make sense right now for the `build_settings` and `toolset` ones):

```python
    def _settings_sha(self):
        if self._adjust_settings_for_std:
            settings = self.settings.copy()
            settings = self._adjust_settings_for_std(settings)
            settings = self._adjust_other_thing(settings)
            settings = self._adjust_...(settings)
            settings = ....
            return settings.sha
        return self.settings.sha
```

Just tell me if you really think it is better to make the copy inside the function and return the sha 😉

File: conans/model/conan_file.py [hunk 12]
```diff
@@ -145,6 +147,10 @@ def initialize(self, settings, env):
         except ConanException:
             pass
 
+        if 'cppstd' in self.settings.fields:
+            self.output.writeln("Setting 'cppstd' is deprecated in favor of 'compiler.cppstd',"
```
Comment (memsharded) [inline conans/model/conan_file.py]: Why not a warning? It will be more visible, and help spotting the message

**Comment by jgsogo** [comment]
Relevant changes since last reviews:
 * **Use `compiler.cppstd` in build_helpers**, until now I wasn't taking into account `compiler.cppstd` in the build helpers 😓. I've written a helper function `cppstd_from_settings`, but not sure if [`conans/client/build/cppstd_flags.py`](https://github.com/conan-io/conan/pull/4917/files#diff-0ae047f956f76dcfd983a59666652cfb) is the place to define it.
 * _Pure C_ recipes will need to remove `compiler.cppstd` (as they were already removing `compiler.libcxx`) or Conan will inject the compiler flag for the C++ standard and some compilers will raise an error
    * So the autogenerated `conanfile.py` for testing has a different checksum.

**Comment by memsharded** [comment]
What happens to existing pure C libraries that are already removing ``compiler.libcxx``, but will not be removing the default added ``compiler.cppstd``? Will the ID be affected?

**Comment by jgsogo** [comment]
> What happens to existing pure C libraries that are already removing `compiler.libcxx`, but will not be removing the default added `compiler.cppstd`? Will the ID be affected?

The problem is not the ID (it won't be affected because the `compiler.cppstd=<default>` is not used to compute the ID), *the problem is that those libraries could fail* if the compiler raises for the _new_ flag... 

image

...and I'm not sure if we can handle this use case, as we can't know that those libraries are _pure C_. Probably this is a **breaking change** ⚠️ ⚠️ ⚠️

**Comment by jgsogo** [comment]
Thinking about this _breaking change_ a little bit deeper:
 * This flag is being introduced in the following elements:
   * `AutoToolsBuildEnvironment`: this _new_ flag goes to `CXXFLAGS`
   * `Meson`: goes to `... -D cpp_std="c++14"`. And that option is different from `c_std` ([docs](http://mesonbuild.com/Builtin-options.html)). 
   * `CMake`: populates:
      ```
      set(CONAN_CMAKE_CXX_STANDARD 14)
      set(CONAN_CMAKE_CXX_EXTENSIONS ON/OFF)
      set(CONAN_STD_CXX_FLAG c++14)
      ...
      set(CMAKE_CXX_FLAGS "${CONAN_STD_CXX_FLAG} ${CMAKE_CXX_FLAGS}")  # CMake <= 3.1
      set(CMAKE_CXX_STANDARD ${CONAN_CMAKE_CXX_STANDARD})
      set(CMAKE_CXX_EXTENSIONS ${CONAN_CMAKE_CXX_EXTENSIONS})
      ```
   * `VisualStudioBuildEnvironment`: goes to dict entry `"CL": "... /std:c++14 ..."` which is then used by `MSBuild` and `VirtualBuildEnvGenerator`.
   * `CompilerArgsGenerator`: it goes directly to the list of flags

So, I think that the problem could be with `CompilerArgsGenerator` (it was the test that failed and I commented before), but all the other ones should avoid this flag when compiling a C library.

File: conans/client/migrations_settings.py [hunk 13]
```diff
@@ -288,3 +288,75 @@
 settings_1_14_2 = settings_1_14_1
 settings_1_14_3 = settings_1_14_2
 settings_1_14_4 = settings_1_14_3
+
+settings_1_15_0 = """
```
Comment (lasote) [inline conans/client/migrations_settings.py]: I think this might be better to be adjusted when the release is merged to develop probably. In that point, we have "freeze" the settings for the version, not before. Why did you add it?

**Comment by jgsogo** [comment]
I need to modify docs according to the _new_ behavior
